dbms/src/Storages/S3/S3GCManager.cpp

```diff
@@ -0,0 +1,495 @@
+        // CheckpointDataFile is a single object, remove it.
+        deleteObject(*client, client->bucket(), datafile_key);
+    }
+    else
+    {
+        // DMFile is composed by multiple objects, need extra work to remove all of them.
+        // TODO: remove all objects belong to this DMFile
+        LOG_WARNING(log, "remove dmfile, key={}", datafile_key);
+    }
+    LOG_INFO(log, "datafile deleted, key={}", datafile_key);
+}
+
+std::vector<UInt64> S3GCManager::getAllStoreIds() const
+{
+    std::vector<UInt64> all_store_ids;
+    // The store key are "s${store_id}/", we need setting delimiter "/" to get the
+    // common prefixes result.
+    // Reference: https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html
+    listPrefix(
+        *client,
+        client->bucket(),
+        /*prefix*/ S3Filename::allStorePrefix(),
+        /*delimiter*/ "/",
+        [&all_store_ids](const Aws::S3::Model::ListObjectsV2Result & result) {
+            const Aws::Vector<Aws::S3::Model::CommonPrefix> & prefixes = result.GetCommonPrefixes();
+            for (const auto & prefix : prefixes)
+            {
+                const auto filename_view = S3FilenameView::fromStoreKeyPrefix(prefix.GetPrefix());
+                RUNTIME_CHECK(filename_view.type == S3FilenameType::StorePrefix, prefix.GetPrefix());
+                all_store_ids.emplace_back(filename_view.store_id);
+            }
+            return PageResult{.num_keys = prefixes.size(), .more = true};
+        });
+
+    return all_store_ids;
+}
+
+std::unordered_set<String> S3GCManager::getValidLocksFromManifest(const String & manifest_key)
+{
+    // download the latest manifest from S3 to local file
+    const String local_manifest_path = getTemporaryDownloadFile(manifest_key);
+    downloadFile(*client, client->bucket(), local_manifest_path, manifest_key);
+    LOG_INFO(log, "Download manifest, from={} to={}", manifest_key, local_manifest_path);
+
+    // parse lock from manifest
+    PS::V3::CheckpointProto::StringsInternMap strings_cache; // TODO: Is there global cache?
+    using ManifestReader = DB::PS::V3::CPManifestFileReader;
+    auto reader = ManifestReader::create(ManifestReader::Options{.file_path = local_manifest_path});
+    auto mf_prefix = reader->readPrefix();
+
+    while (true)
+    {
+        // TODO: calculate the valid size of each CheckpointDataFile in the manifest
+        auto part_edit = reader->readEdits(strings_cache);
+        if (!part_edit)
+            break;
+    }
+
+    std::unordered_set<String> locks;
+    while (true)
+    {
+        auto part_locks = reader->readLocks();
+        if (!part_locks)
+            break;
+        locks.merge(part_locks.value());
+    }
+
+    return locks;
+}
+
+void S3GCManager::removeOutdatedManifest(const CheckpointManifestS3Set & manifests, const Aws::Utils::DateTime * const timepoint)
+{
+    // clean the outdated manifest files
+    auto expired_bound_sec = config.manifest_expired_hour * 3600;
```

Comment:
How about also retaining the last 10 manifests?

Reply:
updated